File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(m, n, puddles) {
    const dp = Array.from({length: n+1}, ()=> Array.from({length: m+1}, ()=> 0));
    const blocked = Array.from({length: n+1}, ()=> Array.from({length: m+1}, ()=>false));
    
    dp[1][1] = 1;
    
    for(const [m,n] of puddles){
        blocked[n][m] = true;
    }
    
    for(let i=1; i<=n; i++){
        for(let j=1; j<=m; j++){
            if(i===1 && j===1)  continue;
            if(blocked[i][j])   continue;
            dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % 1_000_000_007;
        }
    }
    
    return dp[n][m];
}

File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    dp = [[0] * (m+1) for _ in range(n+1)]
    blocked = [[False] * (m+1) for _ in range(n+1)]
    dp[1][1] = 1
    
    for p in puddles:
        x, y = p
        blocked[y][x] = True
    
    for i in range(1, n+1):
        for j in range(1, m+1):
            if i == 1 and j == 1:
                continue
            if blocked[i][j]:
                continue
            dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % 1_000_000_007
    
    return dp[n][m]